### DISCUSSION TITLE
Consider enable mDNS for Zeroconf name service discovery in precompiled Tasmota32 builds for ESP32?

### DISCUSSION THREAD

**Discussion by Hedda** [arendst/Tasmota#15579] [Ideas]
I understand the backstory is that [mDNS was removed/disabled from all pre-compiled binaries of Tasmota to save code space in order to allow new features](https://github.com/arendst/Tasmota/blob/development/CHANGELOG.md#920-20201216), but question if that should also still apply to Tasmota32 for ESP32 which normally have much more flash space available?

https://www.arduino.cc/reference/en/libraries/mdns_generic/

https://github.com/khoih-prog/MDNS_Generic

Request enable mDNS (Multicast DNS Registering Services) for Zeroconf name service discovery precompiled only in Tasmota32 for ESP32?

https://github.com/khoih-prog/MDNS_Generic/blob/master/README.md

I know esp8266_mdns does have a history of not working reliable with Arduino ESP8266, which is why it was disabled in [9.1.0.2] 20201216:

https://github.com/arendst/Tasmota/blob/development/CHANGELOG.md#920-20201216

However, I believe that the mDNS for ESP8266 is not the same code or library as the MDNS_Generic which should be used on ESP32?

https://www.arduino.cc/reference/en/libraries/esp8266_mdns/

https://github.com/mrdunk/esp8266_mdns

MDNS_Generic library looks to be well maintained and seems to get regular bug fixes:

https://github.com/khoih-prog/MDNS_Generic/releases

Hopefully limiting mDNS to Tasmota32 for ESP32 should at least prevent it from causing issues for ESP8266 Tasmota users.

**Comment by Jason2866** [comment]
mDNS will not be enabled in precompiled firmware binaries. mDNS is still buggy in all esp8266 and esp32x frameworks. Since the day we disabled in Tasmota all strange issues where gone away.

**Reply by Hedda** [reply]
Was it by the way "MDNS_Generic by khoih-prog" or another library that was used before when was enabled by default 2-years ago?

https://www.arduino.cc/reference/en/libraries/mdns_generic/

https://github.com/khoih-prog/MDNS_Generic

It was probably not MDNS_Generic by khoih-prog as that only added support for ESP32 and ESP8266 boards in September 2021?

Perhaps the problems with mDNS in Tasmota was in other library, like maybe esp8266_mdns by mrdunk?

https://www.arduino.cc/reference/en/libraries/esp8266_mdns/ 

https://github.com/mrdunk/esp8266_mdns

**Comment by Jason2866** [comment]
Tasmota used the included in Arduino framework.

**Reply by Hedda** [reply]
Is that ArduinoMDNS from arduino-libraries then or?

https://registry.platformio.org/libraries/arduino-libraries/ArduinoMDNS

https://github.com/arduino-libraries/ArduinoMDNS

It does not look like that ArduinoMDNS actively maintained?

Neither does ESP to be specifically listed as supported?

**Comment by Jason2866** [comment]
This one https://github.com/tasmota/Arduino/tree/2.7.4.9/libraries/ESP8266mDNS
and this one https://github.com/tasmota/arduino-esp32/tree/release_203/libraries/ESPmDNS

And there is no time invest planned to try or test any other library.

**Reply by Hedda** [reply]
OK, upstream of those look to get a few updates sometimes, though they do not seem to still be activly developed like those others:

https://github.com/espressif/arduino-esp32/tree/master/libraries/ESPmDNS

https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266mDNS

**Comment by Jason2866** [comment]
Arduino Tasmota Esp32 Core 2.0.3 is up to date with upstream.
Arduino Tasmota Esp8266 Core 2.7.4.9 is NOT up to date with upstream since core 3.0.x does not work reliable for Tasmota (code changes needed?) anyway all functions we need are in our modified core 2.7.4.9.
To be clear we dont care whether mDNS is actual or maintained, since we do not use (by default) nor we recommend to use mDNS. It generates more problems than it solves. 
Tasmota is OpenSource everyone who wants to do changes can do and integrate what she/he ever wants.

Thx, for sharing your idea. To answer your question. mDNS will not be enabled in ANY precompiled Tasmota(32) build.

**Comment by Hedda** [comment]
Any chance you might consider this in the future for a separate prebuilt binary for the Sonoff ZBBridge and Sonoff Zigbee Bridge for users who do not want to use the native [Zigbee2Tasmota](https://tasmota.github.io/docs/Zigbee/) gateway but instead only want to use it as a remote Zigbee Coordinator via serial stream server to access it from other third-party Zigbee gateway software running on a other computer on your local network?

While that is not important to the native [Zigbee2Tasmota](https://tasmota.github.io/docs/Zigbee/) gateway it is a very nice feature for users of Home Assistant's ZHA integration:

https://www.home-assistant.io/integrations/zha#discovery-via-usb-or-zeroconf

That is, when mDNS Zeroconf discovery is enabled it allows Home Assistant to automatically discover it on your local network as a remote Zigbee Coordinator adapter and will show a notification suggesting "Zigbee Home Automation" as a "new devices discovered", the user then has option to just click on configure and so ZHA integration will automatically be installed using DNS record info from mDNS Zeroconf.

Example of direct-attached Sonoff Zigbee dongle USB discovery in Home Assistant which works similar from an end-users point-of-view which the exception that it will show socat socket and mDNS address instead of a local serial device path and USB-ID for the adapter:

![image](https://user-images.githubusercontent.com/6320001/171015940-06992b88-9a51-466b-ba73-42b9b6765dfd.png)

![image](https://user-images.githubusercontent.com/6320001/171015974-1a97d711-1aca-4e56-9ff1-64d71343b979.png)

**Comment by eadmaster** [comment]
i've just tried a custom tasmota32 build based on [this fork](https://github.com/fightforlife/Tasmota/tree/development_newSmartRC), which is working correctly via direct ip connection. 

I've set the `MDNS_ENABLED` to      true in `user_config_override.h`, but mdns does not seem to be working.
I've tried:
 - changing hostname in the Wifi parameters 
 - scanning local network with [`mdns-scan`](https://manpages.ubuntu.com/manpages/focal/man1/mdns-scan.1.html)

**Reply by sfromis** [reply]
As things are now, mDNS support is included in ESP32 standard builds, due to this being needed to support Matter. No need for forks or custom builds, you can just enable this with `SetOption55`.

However, an old fork is "hard to recommend", and is not supported now. Maybe you need:
`#define USE_DISCOVERY`

**Reply by eadmaster** [reply]
> Maybe you need:
> #define USE_DISCOVERY

thank you for the tip, indeed i had `#undef USE_DISCOVERY` in my `user_config_override.h`. I'll try rebuilding and updating if necessary.

UPDATE: rebuild with `USE_DISCOVERY`, mDNS is now working!
